context/src/main/java/io/micronaut/runtime/beans/DefaultIntrospectionBeanMapHandler.java

```diff
@@ -0,0 +1,305 @@
+            mapStrategy = BeanMapper.MapStrategy.DEFAULT;
+        }
+        BeanMapper.MapStrategy.ConflictStrategy conflictStrategy = mapStrategy.conflictStrategy();
+        @NonNull Argument<?>[] constructorArguments = outputIntrospection.getConstructorArguments();
+        if (constructorArguments.length == 0) {
+            O output = outputIntrospection.instantiate();
+            return map(input, output, mapStrategy, outputIntrospection);
+        } else {
+            Objects.requireNonNull(input, "Input cannot be null");
+            // constructor binding
+            Object[] parameters = new Object[constructorArguments.length];
+            for (int i = 0; i < constructorArguments.length; i++) {
+                Argument<?> constructorArgument = constructorArguments[i];
+                String key = constructorArgument.getName();
+                if (input.containsKey(key)) {
+                    Object v = input.get(key);
+                    if (v != null) {
+                        if (constructorArgument.isInstance(v)) {
+                            parameters[i] = v;
+                        } else if (conflictStrategy == BeanMapper.MapStrategy.ConflictStrategy.CONVERT) {
+                            parameters[i] = conversionService.convertRequired(v, constructorArgument);
+                        }
+                    } else if (constructorArgument.isDeclaredNonNull()) {
+                        throw new IllegalArgumentException(PREFIX_NON_NULL_ARG + constructorArgument.getName());
+                    }
+                } else {
+                    throw new IllegalArgumentException("Missing constructor argument from input: " + key);
+                }
+            }
+            O output = outputIntrospection.instantiate(parameters);
+            return map(input, output, mapStrategy, outputIntrospection);
+        }
+    }
+
+    @Override
+    public <O> O map(Map<String, Object> input, O output, BeanMapper.MapStrategy mapStrategy, BeanIntrospection<O> right) {
+        if (mapStrategy == null) {
+            mapStrategy = BeanMapper.MapStrategy.DEFAULT;
+        }
+        BeanMapper.MapStrategy.ConflictStrategy conflictStrategy = mapStrategy.conflictStrategy();
+        if (CollectionUtils.isNotEmpty(input)) {
+            boolean shouldError = conflictStrategy == null || conflictStrategy == BeanMapper.MapStrategy.ConflictStrategy.ERROR;
+            input.forEach((key, value) -> {
+                boolean isNonNull = value != null;
+                if (shouldError) {
+                    mapOrError(output, right, key, value, isNonNull);
+                } else {
+                    mapEntry(output, conflictStrategy, right, key, value, isNonNull);
+                }
+            });
+        }
+        return output;
+    }
+
+    private static <O> void mapEntry(O output, BeanMapper.MapStrategy.ConflictStrategy conflictStrategy, BeanIntrospection<O> right, String key, Object value, boolean isNonNull) {
```

Comment:
We might want to support `with` like methods as well

Reply:
Agreed